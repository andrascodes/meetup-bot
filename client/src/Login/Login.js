import React, { Component } from 'react'
import { Alert } from 'react-bootstrap'

import './Login.css'

import { createAuthApi } from '../lib/authApi'

export class Login extends Component {

  state = {
    username: '',
    password: '',
    error: false,
    alertVisible: false
  }

  login = createAuthApi(fetch).login

  handleUsernameChange = e => {
    e.persist()
    e.preventDefault()

    this.setState(state => ({
      username: e.target.value
    }))
  }
  
  handlePasswordChange = e => {
    e.persist()
    e.preventDefault()

    this.setState(state => ({
      password: e.target.value
    }))
  }

  handleSubmit = e => {
    e.persist()
    e.preventDefault()

    this.login(this.state.username, this.state.password)
    .then(({ status, token, body }) => {
      localStorage.apiToken = token
      localStorage.username = body.username
      localStorage.email = body.email

      let redirectTo = this.props.location
      while(redirectTo.state) {
        redirectTo = redirectTo.state.from
      }
      redirectTo = redirectTo.pathname

      this.props.history.push(redirectTo === '/login' || redirectTo === '/login/' ? '/' : redirectTo)
    })
    .catch(error => {
      console.error(error)
      this.setState(state => ({ error, alertVisible: true }))
    })
  }

  handleErrorDismiss = e => {
    e.persist()
    e.preventDefault()

    this.setState(state => ({
      alertVisible: false
    }))
  }

  render() {
    return (
      <div className="Login container">
        <div className="row">
          <div className="col-md-4 col-md-offset-4">
            <div className="login-panel panel panel-default">
              
              <div className="panel-heading">
                <h3 className="panel-title">
                  Please sign in
                </h3>
              </div>

              <div className="panel-body">
                <form role="form" onSubmit={this.handleSubmit}>
                  <fieldset>
                    <div className="form-group">
                      <label htmlFor="usernameInput">Username</label>
                      <input 
                        type="text" 
                        className="form-control" 
                        id="usernameInput" 
                        placeholder="Username" 
                        onChange={this.handleUsernameChange}
                      />
                    </div>
                    
                    <div className="form-group">
                      <label htmlFor="passwordInput">Password</label>
                      <input 
                        type="password" 
                        className="form-control" 
                        id="passwordInput" 
                        placeholder="Password" 
                        onChange={this.handlePasswordChange}
                      />
                    </div>
                    
                    <button type="submit" className="btn btn-default">Submit</button>
                  
                  </fieldset>
                </form>
              </div>

            </div>
          </div>
        </div>

        <div className="row">
          <div className="col-md-4 col-md-offset-4">
            {
              this.state.alertVisible ? 
              <Alert bsStyle="danger" onDismiss={this.handleErrorDismiss}>
                <h4>{this.state.error.error || this.state.error.toString()}</h4>
              </Alert>
              :
              ''
            }
          </div>
        </div>
      </div>
    )
  }
}