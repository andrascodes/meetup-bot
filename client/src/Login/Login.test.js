import React from 'react'
// import { MemoryRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import { Login } from './Login'

describe(`Login`, () => {

  beforeEach(() => {
    global.localStorage = {

    }
  })

  test('renders correctly', () => {
    const wrapper = mount(<Login />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('renders when state.username is changed', () => {
    const wrapper = mount(<Login />)
   
    const username = 'user_1'
    wrapper.find('#usernameInput').simulate('change', { 
      persist: () => {}, 
      preventDefault: () => {},
      target: {
        value: username,
      }
    })
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('renders when state.password is changed', () => {
    const wrapper = mount(<Login />)
   
    const password = 'another_password'
    wrapper.find('#passwordInput').simulate('change', { 
      persist: () => {}, 
      preventDefault: () => {},
      target: {
        value: password,
      }
    })
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  
  test('renders when state.error is set', () => {
    const wrapper = mount(<Login />)
    
    wrapper.setState({
      error: {
        error: 'Username is incorrect.'
      },
      alertVisible: true
    })
    expect(toJson(wrapper)).toMatchSnapshot()

    wrapper.setState({
      error: {
        toString: () => 'Error: Username is incorrect.'
      },
      alertVisible: true
    })
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('closes the Error Alert panel when close button is clicked', () => {
    const wrapper = mount(<Login />)

    wrapper.setState({
      error: {
        toString: () => 'Error: Username is incorrect.'
      },
      alertVisible: true
    })

    wrapper.find('#errorClose > .close').find('[aria-hidden="true"]').simulate('click', { 
      persist: () => {}, 
      preventDefault: () => {},
    })
    expect(toJson(wrapper)).toMatchSnapshot()
  })

})