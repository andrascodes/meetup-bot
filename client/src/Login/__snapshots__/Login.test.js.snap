exports[`Login closes the Error Alert panel when close button is clicked 1`] = `
<Login>
  <div
    className="Login container">
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <div
          className="login-panel panel panel-default">
          <div
            className="panel-heading">
            <h3
              className="panel-title">
              Please sign in
            </h3>
          </div>
          <div
            className="panel-body">
            <form
              onSubmit={[Function]}
              role="form">
              <fieldset>
                <div
                  className="form-group">
                  <label
                    htmlFor="usernameInput">
                    Username
                  </label>
                  <input
                    className="form-control"
                    id="usernameInput"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text" />
                </div>
                <div
                  className="form-group">
                  <label
                    htmlFor="passwordInput">
                    Password
                  </label>
                  <input
                    className="form-control"
                    id="passwordInput"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password" />
                </div>
                <button
                  className="btn btn-default"
                  type="submit">
                  Submit
                </button>
              </fieldset>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4" />
    </div>
  </div>
</Login>
`;

exports[`Login renders correctly 1`] = `
<Login>
  <div
    className="Login container">
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <div
          className="login-panel panel panel-default">
          <div
            className="panel-heading">
            <h3
              className="panel-title">
              Please sign in
            </h3>
          </div>
          <div
            className="panel-body">
            <form
              onSubmit={[Function]}
              role="form">
              <fieldset>
                <div
                  className="form-group">
                  <label
                    htmlFor="usernameInput">
                    Username
                  </label>
                  <input
                    className="form-control"
                    id="usernameInput"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text" />
                </div>
                <div
                  className="form-group">
                  <label
                    htmlFor="passwordInput">
                    Password
                  </label>
                  <input
                    className="form-control"
                    id="passwordInput"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password" />
                </div>
                <button
                  className="btn btn-default"
                  type="submit">
                  Submit
                </button>
              </fieldset>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4" />
    </div>
  </div>
</Login>
`;

exports[`Login renders when state.error is set 1`] = `
<Login>
  <div
    className="Login container">
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <div
          className="login-panel panel panel-default">
          <div
            className="panel-heading">
            <h3
              className="panel-title">
              Please sign in
            </h3>
          </div>
          <div
            className="panel-body">
            <form
              onSubmit={[Function]}
              role="form">
              <fieldset>
                <div
                  className="form-group">
                  <label
                    htmlFor="usernameInput">
                    Username
                  </label>
                  <input
                    className="form-control"
                    id="usernameInput"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text" />
                </div>
                <div
                  className="form-group">
                  <label
                    htmlFor="passwordInput">
                    Password
                  </label>
                  <input
                    className="form-control"
                    id="passwordInput"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password" />
                </div>
                <button
                  className="btn btn-default"
                  type="submit">
                  Submit
                </button>
              </fieldset>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <Alert
          bsClass="alert"
          bsStyle="danger"
          closeLabel="Close alert"
          id="errorClose"
          onDismiss={[Function]}>
          <div
            className="alert alert-danger alert-dismissable"
            id="errorClose"
            role="alert">
            <button
              aria-hidden="true"
              className="close"
              onClick={[Function]}
              tabIndex="-1"
              type="button">
              <span>
                ×
              </span>
            </button>
            <h4>
              Username is incorrect.
            </h4>
            <button
              className="close sr-only"
              onClick={[Function]}
              type="button">
              Close alert
            </button>
          </div>
        </Alert>
      </div>
    </div>
  </div>
</Login>
`;

exports[`Login renders when state.error is set 2`] = `
<Login>
  <div
    className="Login container">
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <div
          className="login-panel panel panel-default">
          <div
            className="panel-heading">
            <h3
              className="panel-title">
              Please sign in
            </h3>
          </div>
          <div
            className="panel-body">
            <form
              onSubmit={[Function]}
              role="form">
              <fieldset>
                <div
                  className="form-group">
                  <label
                    htmlFor="usernameInput">
                    Username
                  </label>
                  <input
                    className="form-control"
                    id="usernameInput"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text" />
                </div>
                <div
                  className="form-group">
                  <label
                    htmlFor="passwordInput">
                    Password
                  </label>
                  <input
                    className="form-control"
                    id="passwordInput"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password" />
                </div>
                <button
                  className="btn btn-default"
                  type="submit">
                  Submit
                </button>
              </fieldset>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <Alert
          bsClass="alert"
          bsStyle="danger"
          closeLabel="Close alert"
          id="errorClose"
          onDismiss={[Function]}>
          <div
            className="alert alert-danger alert-dismissable"
            id="errorClose"
            role="alert">
            <button
              aria-hidden="true"
              className="close"
              onClick={[Function]}
              tabIndex="-1"
              type="button">
              <span>
                ×
              </span>
            </button>
            <h4>
              Error: Username is incorrect.
            </h4>
            <button
              className="close sr-only"
              onClick={[Function]}
              type="button">
              Close alert
            </button>
          </div>
        </Alert>
      </div>
    </div>
  </div>
</Login>
`;

exports[`Login renders when state.password is changed 1`] = `
<Login>
  <div
    className="Login container">
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <div
          className="login-panel panel panel-default">
          <div
            className="panel-heading">
            <h3
              className="panel-title">
              Please sign in
            </h3>
          </div>
          <div
            className="panel-body">
            <form
              onSubmit={[Function]}
              role="form">
              <fieldset>
                <div
                  className="form-group">
                  <label
                    htmlFor="usernameInput">
                    Username
                  </label>
                  <input
                    className="form-control"
                    id="usernameInput"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text" />
                </div>
                <div
                  className="form-group">
                  <label
                    htmlFor="passwordInput">
                    Password
                  </label>
                  <input
                    className="form-control"
                    id="passwordInput"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password" />
                </div>
                <button
                  className="btn btn-default"
                  type="submit">
                  Submit
                </button>
              </fieldset>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4" />
    </div>
  </div>
</Login>
`;

exports[`Login renders when state.username is changed 1`] = `
<Login>
  <div
    className="Login container">
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4">
        <div
          className="login-panel panel panel-default">
          <div
            className="panel-heading">
            <h3
              className="panel-title">
              Please sign in
            </h3>
          </div>
          <div
            className="panel-body">
            <form
              onSubmit={[Function]}
              role="form">
              <fieldset>
                <div
                  className="form-group">
                  <label
                    htmlFor="usernameInput">
                    Username
                  </label>
                  <input
                    className="form-control"
                    id="usernameInput"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text" />
                </div>
                <div
                  className="form-group">
                  <label
                    htmlFor="passwordInput">
                    Password
                  </label>
                  <input
                    className="form-control"
                    id="passwordInput"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password" />
                </div>
                <button
                  className="btn btn-default"
                  type="submit">
                  Submit
                </button>
              </fieldset>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row">
      <div
        className="col-md-4 col-md-offset-4" />
    </div>
  </div>
</Login>
`;
